File: R/SVM.R
library(e1071)
library(bench)
setwd("/Users/ancavanciupopescu/Desktop/Classes/CS 32/Final project/Data/")
set.seed(32)
data <- read.csv("SVM_data.csv")[,-1]
y <- data[,1]
X <- as.matrix(data[,2:20])

# SVM using e1071 package
svm_package <- function(X, y) {
  model <- svm(x = X, y = as.factor(y), kernel = "linear", type = "C-classification", scale = FALSE)
  list(coefficients = t(as.vector(model$coefs) %*% model$SV), intercept = -model$rho)
}

# Simple linear SVM from scratch
svm_from_scratch <- function(X, y, epochs = 100, learning_rate = 0.01, lambda = 0.01) {
  n_samples <- nrow(X)
  n_features <- ncol(X)
  weights <- matrix(0, nrow = n_features, ncol = 1)
  intercept <- 0
  
  for (epoch in 1:epochs) {
    for (i in 1:n_samples) {
      if (y[i] * (crossprod(weights, X[i, ]) + intercept) < 1) {
        weights <- weights + learning_rate * ((y[i] * matrix(X[i, ], ncol = 1)) - (2 * lambda * weights))
        intercept <- intercept + learning_rate * y[i]
      } else {
        weights <- weights - learning_rate * (2 * lambda * weights)
      }
    }
  }
  list(coefficients = weights, intercept = intercept)
}

# Measure execution time and memory
results <- bench::mark(
  package = { package_result <- svm_package(X, y) },
  scratch = { scratch_result <- svm_from_scratch(X, y) },
  iterations = 10,
  check = FALSE
)
cat("Execution time for e1071:", results$total_time[1], "\n")
cat("Memory allocation for e1071:", results$mem_alloc[1]/1e6, "\n")
cat("Coefficients from package:\n", package_result$coefficients, "\nIntercept from package:", package_result$intercept, "\n")
cat("Execution time from scratch:", results$total_time[2], "\n")
cat("Memory allocation from scratch:", results$mem_alloc[2]/1e6, "\n")
cat("Coefficients from scratch:\n", scratch_result$coefficients, "\nIntercept from scratch:", scratch_result$intercept, "\n")

File: Python/SVM.py
import numpy as np
from sklearn.svm import SVC
from memory_profiler import profile
import time
import pandas as pd

# Load data
path = "/Users/ancavanciupopescu/Desktop/Classes/CS 32/Final project/Data/SVM_data.csv"
data = pd.read_csv(path).iloc[:, 1:]
y = data.iloc[:, 0].values
X = data.iloc[:, 1:].values

# SVM from scratch
@profile
def svm_from_scratch(X, y, epochs=500, learning_rate=0.01, C=1.0):
    w = np.zeros(X.shape[1])
    b = 0
    for epoch in range(epochs):
        for i in range(len(y)):
            if y[i] * (np.dot(X[i], w) + b) < 1:
                w += learning_rate * ((y[i] * X[i]) + (-2 * (1/C) * w))
                b += learning_rate * y[i]
            else:
                w -= learning_rate * (2 * (1/C) * w)
    return w, b

# Scikit-Learn SVM
@profile
def svm_scikit_learn(X, y):
    clf = SVC(kernel='linear', C=1.0)
    clf.fit(X, y)
    return clf.coef_[0], clf.intercept_[0]

# Measure execution and memory
start_time_scratch = time.time()
weights_scratch, bias_scratch = svm_from_scratch(X, y)
end_time_scratch = time.time()
scratch_time = end_time_scratch - start_time_scratch
start_time_skl = time.time()
weights_skl, bias_skl = svm_scikit_learn(X, y)
end_time_skl = time.time()
skl_time = end_time_skl - start_time_skl
print("Weights and bias from scratch:", weights_scratch, bias_scratch)
print("Execution time for SVM from scratch:", scratch_time, "seconds")
print("Weights and bias from scikit-learn:", weights_skl, bias_skl)
print("Execution time for SVM with scikit-learn:", skl_time, "seconds")